master/internal/cache/cache.go

```diff
@@ -0,0 +1,199 @@
+		caches:  make(map[int]*modelDefFolder),
+	}
+}
+
+func (f *FileCache) getOrCreateFolder(expID int) (*modelDefFolder, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	value, ok := f.caches[expID]
+	if ok {
+		return value, nil
+	}
+
+	exp := struct {
+		ModelDefinition []byte
+	}{}
+	err := db.Bun().NewSelect().TableExpr(
+		"experiments").Column("model_definition").Where("id = ?", expID).Scan(context.Background(), &exp)
+	if err != nil {
+		return nil, err
+	}
+	var fileTree []*experimentv1.FileNode
+	arc, err := archive.FromTarGz(exp.ModelDefinition)
+	if err != nil {
+		return nil, err
+	}
+	err = os.MkdirAll(f.genPath(expID, ""), 0700)
+	if err != nil {
+		return nil, err
+	}
+	for _, ar := range arc {
+		if ar.IsDir() {
+			err = os.MkdirAll(f.genPath(expID, ar.Path), 0700)
+		} else {
+			err = os.WriteFile(f.genPath(expID, ar.Path), ar.Content, 0600)
+		}
+		if err != nil {
+			return nil, err
+		}
+		fileTree = append(fileTree, &experimentv1.FileNode{
+			Path:          ar.Path,
+			ModifiedTime:  timestamppb.New(ar.ModifiedTime.Time),
+			ContentLength: int32(len(ar.Content)),
+			IsDir:         ar.IsDir(),
+			ContentType:   http.DetectContentType(ar.Content),
+			Name:          filepath.Base(ar.Path),
+		})
+	}
+	value = &modelDefFolder{
+		path:       f.genPath(expID, ""),
+		fileTree:   fileTree,
+		cachedTime: time.Now(),
+	}
+	f.caches[expID] = value
+	f.prune()
+
+	return value, nil
+}
+
+func (f *FileCache) prune() {
+	for expID, folder := range f.caches {
+		if folder.cachedTime.Add(f.maxAge).Before(time.Now()) {
+			err := os.RemoveAll(folder.path)
+			if err != nil {
+				log.WithError(err).Errorf("failed to prune model definition cache under %s", folder.path)
+			}
+			delete(f.caches, expID)
+		}
+	}
+}
+
+func (f *FileCache) genPath(expID int, path string) string {
+	return filepath.Join(f.rootDir, strconv.Itoa(expID), path)
+}
+
+func (f *FileCache) getFileTree(expID int) ([]*experimentv1.FileNode, *modelDefFolder, error) {
+	folder, err := f.getOrCreateFolder(expID)
+	if err != nil {
+		return nil, nil, err
+	}
+	folder.lock.RLock()
+	defer folder.lock.RUnlock()
+	return folder.fileTree, folder, nil
+}
+
+// GetFileTreeNested returns folder tree structure with given experiment id.
+func (f *FileCache) GetFileTreeNested(expID int) ([]*experimentv1.FileNode, error) {
+	fileTree, _, err := f.getFileTree(expID)
+	if err != nil {
+		return nil, err
+	}
+	return genNestedTree(fileTree), nil
+}
+
+// GetFileContent returns file with given experiment id and path.
+func (f *FileCache) GetFileContent(expID int, path string) ([]byte, error) {
+	fileTree, folder, err := f.getFileTree(expID)
+	if err != nil {
+		return []byte{}, err
+	}
+	for _, file := range fileTree {
+		if file.Path == path {
+			folder.lock.Lock()
+			defer folder.lock.Unlock()
+			file, err := os.ReadFile(f.genPath(expID, path))
+			if err != nil {
+				_, ok := err.(*fs.PathError)
+				if ok {
+					log.Debugf(`File system cache (%s) is likely out of sync. 
```

Comment:
Please log at error level.  If this shows up in a user's master log, I want it to be obvious that something is going wrong.

Reply:
Makes sense, this is fixed